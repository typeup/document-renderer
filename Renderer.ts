import * as dom from "@typeup/dom"
export type renderFunction = (renderer: Renderer, node: dom.Node) => Promise<string>

export class Renderer {
	private constructor(private variables: { [name: string]: string } = {}) {
	}
	getVariable(name: string): string {
		return this.variables[name]
	}
	setVariable(name: string, value: string) {
		this.variables[name] = value
	}
	async render(node: dom.Node | dom.Node[]): Promise<string> {
		let result: string
		if (Array.isArray(node))
			result = (await Promise.all(node.map(n => this.render(n)))).reduce<string[]>((r, e) => Array.isArray(e) ? [...r, ...e] : [...r, e], []).join("")
		else {
			const render = renderers[node.class]
			result = render ? await render(this, node) : ""
		}
		return result
	}
	scope(): Renderer {
		return new Renderer({ ...this.variables })
	}
	static create(): Renderer {
		return new Renderer({})
	}
}
const renderers: { [className: string]: renderFunction } = {}
export function addRenderer(className: string, render: renderFunction) {
	renderers[className] = render
}
