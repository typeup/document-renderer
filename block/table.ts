import * as dom from "@typeup/dom"
import { Renderer, addRenderer } from "../Renderer"

function getClasses(me: dom.block.Table): string {
	return me.alignments.map((value, index) => {
		let result: string
		switch (value) {
			case "center":
				result = "tc" + (index + 1)
				break
			case "right":
				result = "tr" + (index + 1)
				break
			default:
				result = ""
				break
		}
		return result
	}).filter(value => value != "").join(" ")
}

async function render(renderer: Renderer, me: dom.block.Table): Promise<string> {
	return `<table class="${ getClasses(me) }">
	${ (await Promise.all(me.rows.map(async row => renderer.render(row)))).join("") }
	<caption>${ await renderer.render(me.content) }</caption>
</table>`
}
addRenderer("Block.Table", render)
