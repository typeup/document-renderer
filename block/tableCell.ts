import * as dom from "@typeup/dom"
import { Renderer, addRenderer } from "../Renderer"

async function render(renderer: Renderer, me: dom.block.TableCell): Promise<string> {
	return `<t${ me.header ? "h" : "d" }>
	${ await renderer.render(me.content) }
</t${ me.header ? "h" : "d" }>`
}
addRenderer("Block.TableCell", render)
