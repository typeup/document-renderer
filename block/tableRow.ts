import * as dom from "@typeup/dom"
import { Renderer, addRenderer } from "../Renderer"

async function render(renderer: Renderer, me: dom.block.TableRow): Promise<string> {
	return `<tr>
	${ (await Promise.all(me.content.map(async row => renderer.render(row)))).join("") }
</tr>`
}
addRenderer("Block.TableRow", render)
