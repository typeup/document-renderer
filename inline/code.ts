import * as dom from "@typeup/dom"
import { Renderer, addRenderer } from "../Renderer"

async function render(renderer: Renderer, me: dom.inline.Code): Promise<string> {
	return `<code>${me.value.replace(/\&/gi, "&amp;").replace(/</gi, "&lt;").replace(/>/gi, "&gt;")}</code>`
}
addRenderer("Inline.Code", render)
