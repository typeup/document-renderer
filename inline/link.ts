import * as dom from "@typeup/dom"
import { Renderer, addRenderer } from "../Renderer"

async function render(renderer: Renderer, me: dom.inline.Link): Promise<string> {
	return `<a href='${me.target}'>${ await renderer.render(me.content) }</a>`
}
addRenderer("Inline.Link", render)
