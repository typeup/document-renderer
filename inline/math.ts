import * as dom from "@typeup/dom"
import { Renderer, addRenderer } from "../Renderer"

async function render(renderer: Renderer, me: dom.inline.Math): Promise<string> {
	return `$${ me.value }$`
}
addRenderer("Inline.Math", render)
