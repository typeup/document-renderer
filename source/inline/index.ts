import "./code"
import "./emphasize"
import "./index"
import "./link"
import "./math"
import "./text"
