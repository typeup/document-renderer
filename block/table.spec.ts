import * as dom from "@typeup/dom"
import { Renderer } from "../Renderer"
import "../inline"
import "./tableCell"
import "./tableRow"
import "./table"

const renderer = Renderer.create()
describe("Block.Table", () => {
	it("render", async () => {
		expect(await renderer.render(new dom.block.Table(["", ""], [
			new dom.block.TableRow([
				new dom.block.TableCell(true, [new dom.inline.Text("Header A")]),
				new dom.block.TableCell(true, [new dom.inline.Text("Header B")]),
			]),
			new dom.block.TableRow([
				new dom.block.TableCell(false, [new dom.inline.Text("value A")]),
				new dom.block.TableCell(false, [new dom.inline.Text("value B")]),
			]),
		], 
		[
			new dom.inline.Text("Table caption.")
		]))).toMatchSnapshot()
	})
})
