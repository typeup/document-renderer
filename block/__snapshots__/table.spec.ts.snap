// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Block.Table render 1`] = `
"<table class=\\"\\">
	<tr>
	<th>
	Header A
</th><th>
	Header B
</th>
</tr><tr>
	<td>
	value A
</td><td>
	value B
</td>
</tr>
	<caption>Table caption.</caption>
</table>"
`;
